### ISSUE TITLE
API-9-Support Java 11 in OpenLiberty

### ISSUE THREAD

**Issue by aguibert** [OpenLiberty/open-liberty#540]
Deliver Java 11 support to OpenLiberty so that it can be run on Java 7, 8, or 11. 

OpenLiberty should "work with" Java 9, and 10, but since these will not be long-term support releases, we will instead focus our efforts ahead on Java 11.

Significant changes in Java 11:
1. JDK modularization: The JDK itself (and eventually software up the stack) is being converted to modules that will restrict visibility both at compile time and runtime. [1] [2]
1. XML impl change: The Oracle JDK is changing to use a new XML implementation, and the IBM JDK will inherit this change as well [1]
1. IBM JDK Security impl change: The IBM JDK is getting rid of its core security impl and will use the same security impl that the Oracle JDK uses. Note that some IBM JDK security related value-adds will still be available. [1]

[1]: IBM JDK 9 compatibility documentation
https://www.ibm.com/support/knowledgecenter/SSYKE2_9.0.0/com.ibm.java.multiplatform.90.doc/user/compatibility.html
[2]: JEP 261 - Java Platform Module System
http://openjdk.java.net/jeps/261

**Comment by aguibert** [comment]
As of 10/13/2017, we have 55,078/55,645 tests (unit + functional) passing.

Here is a list of which functional areas are passing on Java 9:
- [ ] application management
- [ ] application client
- [ ] beanvalidation
- [ ] CDI (6 tests failing)
- [ ] channel framework
- [ ] classloading
- [ ] concurrency, EE
- [ ] config (1 test failing)
- [ ] EJB container (2 tests failing for remote EJB)
- [ ] expression language 3.0
- [ ] javamail
- [ ] JAX-RS 2.0 (6 tests failing)
- [ ] JAX-RS 2.1 (6 tests failing)
- [ ] JAX-WS (63 tests failing)
- [ ] jbatch 
- [ ] JCA
- [ ] JDBC (1 test failing)
- [ ] JNDI (6 tests failing)
- [ ] JPA
- [ ] JSF
- [ ] JSON-B
- [ ] JSON-P
- [ ] JSP
- [ ] Kernel (6 tests failing)
- [ ] management (J2EE)
- [ ] JMS
- [ ] MicroProfile Config
- [ ] MicroProfile Fault Tolerance
- [ ] MicroProfile Health
- [ ] OpenAPI
- [ ] Security (26 tests failing)
- [ ] SSL
- [ ] Threading
- [ ] Transactions
- [ ] Transport (IIOP + HTTP)
- [ ] Web Container (3 tests failing)

**Comment by gireeshpunathil** [comment]
@aguibert - Is this a work in progress? Fixing ~600 test failures seem too much work, do you think some collaboration will help move things fast? I can spare some bandwidth, please let me know.

**Comment by aguibert** [comment]
@gireeshpunathil yes this is still a work in progress, and has been until late 2015.  I don't expect that the remaining failures are all distinct issues, probably more like 10-20 unique issues.  For comparison, when I first started this effort we had >20k failures, so we've come a long way just to reach this point.  

Thank you for offering to help on this!  Unfortunately, many of the failing functional tests have not been made public yet (the team is actively working to bring things over into the open), so helping out with the debugging of these particular failing tests will not be possible from your position.  If anything comes up I will post back to this issue with ways the larger community can help out.

**Comment by aguibert** [comment]
Since Java 9 (and 10) will not have long-term support, I am repurposing this item for Java 11 support, which is the next LTS release of Java.  Of course, we should still do our best to work with Java 9 and 10, and will use testing with these releases as stepping stones to Java 11 support.

**Comment by aguibert** [comment]
Moving on to OpenJ9 + OpenJDK 10 we have decent test results.  Using a nightly build from 2018-06-11, we have 60463/60708 tests passing (99.6% pass rate).

At this point I'm going to move on to the early access Hotspot + OpenJDK 11 builds and see if I can get OpenLiberty to build fully.

**Comment by ChengJin01** [comment]
@aguibert ,  I am wondering whether there is any latest internal build (supporting Java 11) for downloading. We need to use it to do perf test with Java 11 if available.

**Comment by aguibert** [comment]
hi @ChengJin01, I've been testing OpenLiberty nightly builds with OpenJDK 11 build 28 (http://jdk.java.net/11/)

You can download the latest internal OpenLiberty build here: https://openliberty.io/downloads/
(scroll down to the "development builds" section)

**Comment by aguibert** [comment]
Current test status for OpenLiberty w/ OpenJDK 11:
62,605 tests running, 1,644 failing (97.4% passrate)

The main problem areas we are currently working through:
 - accommodating for removed JDK APIs (e.g. JAX-B and CORBA)
 - issues with TLSv1.3 (new protocol introduced in JDK 11)
 - TLSv1.3 not supporting DSA signed certificates, which many of our automated tests use as dummy self-signed certificates

**Comment by atosak** [comment]
Ship readiness checklist is being tracked here:  https://github.com/OpenLiberty/open-liberty/issues/5450

**Comment by atosak** [comment]
Ship readiness checklist is being tracked here: https://github.com/OpenLiberty/open-liberty/issues/5450

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same BOX folder.  Thanks!

**Comment by aguibert** [comment]
OpenLiberty is now officially supported on Java 11! 🎉 🎉 🎉 

I posted a blog about what's new and where to get Java 11 here: https://openliberty.io/blog/2019/02/06/java-11.html

Will close out this epic once I get the approval for my newly created test summary by @sabolo.

**Comment by sabolo** [comment]
At present I cannot close out this feature.  While we now have support for OpenJ9 Java 11, HotSpot still has not been delivered.  The feature, in its current form, is for supporting all Java 11 and I am not aware of any other features for HotSpot support so as a result we need to keep this open at this time.  From an OpenJ9 perspective I am happy.

**Comment by marikaj123** [comment]
The bulk of the work went in 19001. Andy will provide at least 1 PR to enable TLSv13 for hotspot, but probably more for test updates.  Thereafter, Kevin will test and provide FAT approval.

**Comment by marikaj123** [comment]
According to the development team, the feature was partially complete in 19001.

**Comment by lostiniceland** [comment]
Using Openliberty 19.0.0.2 with OpenJ9 I get errors from Eclipselinks validation because of a Java-Enum used as a property in an entity (Eclipselink complains it is not serializable, even though I am using a JPA 2.2 converter).

Fun thing is, switching back to Java 8, Eclipselink runs fine without any errors or warnings.

It is really just a simple plain entity with a couple of fields and one of them is a enum.

**Comment by aguibert** [comment]
hi @lostiniceland, thanks for trying this out and reporting the issue. We have thousands of integration tests that we ran with JPA and Java 11, so I'm surprised this wasn't caught yet. Can you please open a separate issue so we can track it and discuss further?

**Comment by lostiniceland** [comment]
@aguibert sure thing. I will try to provide a little sample application with the issue

**Comment by marlowa** [comment]
Is there an ETA for the enum issue reported by lostiniceland? 
I am trying to move a project from JDK8 to JDK11 but it uses WebSphere. While at QCON I spoke to someone from IBM during a talk about migrating from JDK8 to JDK11. When I told him I was using WebSphere he said that Liberty was ready but from here it looks like it isn't quite ready yet. I am concerned.

**Comment by lostiniceland** [comment]
I havent been able to report the additional issue yet (probably next weekend). I've also found a problem when using an dedicated Converter for enums (mapping enum to character) where a JPQL is broken under Java 11 (which works on 8).
I assume all those issues need to be fixed in Eclipselink though.

**Comment by jgrassel** [comment]
@lostiniceland Eclipselink 2.7.4 (jpa-2.2 feature) has been updated to better work with JDK 11 with the following commit:

commit 2c008901517766ee1b18000698b3466e786494b6
Merge: 9b63c960a b13f3213e
Author: Will Dazey <dazeydev.3@gmail.com>
Date:   Fri Mar 8 10:18:28 2019 -0600

    Merge pull request #6268 from dazey3/6266_integration
    
    Issue 6266: Update EclipseLink from 2.7.3 to 2.7.4 for JPA 2.2

**Comment by dazey3** [comment]
@lostiniceland This update from EclipseLink 2.7.3 -> 2.7.4 adds Java 12 support to EclipseLink. However, I would suggest testing against this latest build of EclipseLink in case the issue with Converters has been fixed. If it has not, I would be very interested in looking at a sample so that I can investigate this issue. Also, have you checked EclipseLink's bug tracker to see if there is an existing, open issue matching this failure? https://bugs.eclipse.org/bugs/query.cgi (EclipseLink is a product under EE4J)

**Comment by lostiniceland** [comment]
I was able to reproduce all (currently) three issues with a sample project. Only Java 11 WARs compiled with Java 11 are affected. Compiling the same project with Java 1.8 and running on a Java 11 Openliberty everything works. 
Please see issue #6948

**Comment by aguibert** [comment]
We should be able to support JDK 11 on Hotspot once OpenJDK 11.0.3 is released (ETA April 18th), which contains some TLSv13 fixes that we were able to port to OpenJ9 earlier.

**Comment by aguibert** [comment]
Support for TLSv1.3 has been turned on for OpenJDK 11 w/ Hotspot, meaning that OpenLiberty will now work with both of the main JVMs (OpenJ9 and Hotspot).

As long as nothing moves backwards at this point, OpenJDK 11 w/ Hotspot support will be available in 19005.
